server/src/main/java/org/elasticsearch/env/Environment.java

```diff
@@ -156,7 +158,12 @@ public Environment(final Settings settings, final Path configPath) {
             assert sharedDataFile != null;
             finalSettings.put(Environment.PATH_SHARED_DATA_SETTING.getKey(), sharedDataFile.toString());
         }
-        this.settings = finalSettings.build();
+
+        if (DiscoveryNode.isStateless(settings)) {
+            this.settings = StatelessSecureSettings.install(settings);
```

Comment:
Sorry I missed this before. We do actually need the final settings here, they are currently lost. Since you are setting the secure settings on the builder inside the install method, I would pass in the builder (to set secure settings on), and settings (as the source) separately.

Reply:
Hi @rjernst , I tried what you said but unfortunately it seems not possible because there's already some sort of secure settings set on the builder:

```
[2022-12-09T19:41:31,344][ERROR][o.e.b.Elasticsearch      ] [kingherc-ub] fatal exception while booting Elasticsearchjava.lang.IllegalArgumentException: Secure settings already set. Existing settings: [keystore.seed], new settings: [s3.client.default.secret_key, s3.client.default.access_key, s3.client.default.session_token]
```

I reverted this line back to accepting `finalSettings.build()` instead of just `settings`. Even though we're re-building the settings, it seems OK to me since this is a temporary measure. But feel free to review and tell me if it's important enough to chase down where the original secure settings are set and avoid them.